src/Network/GRPC/MQTT/Message/Packet.hs

```diff
@@ -371,4 +412,111 @@ isMissingPackets = do
+-- over memory is desired. This function is significantly more unsafe when 
+-- compared to 'wireWrapPacket', which automatically handles allocation of the 
+-- serialization buffer internally, since it assumes that the @ptr@ refers to a 
+-- live allocation and the allocation size @len@ always equal to or larger than 
+-- the size of the serialized @packet@.
+--
+-- @since 1.0.0
+encodeBufferOffPacket ::
+  MonadIO m =>
+  -- | The @ptr@ argument is pointer to the start of the proto serialization 
+  -- buffer. This pointer refer to a pinned 'MutableByteArray' aligned to 
+  -- 'metaDataAlign'.
+  Ptr Word8 ->
+  -- | The @len@ argument is the size of the buffer @ptr@ points to. The value 
+  -- of length must exactly agree with the actual size of the allocation and 
+  -- must be large enough to store the serialized @packet@ plus 'metaDataSize'.
+  Int ->
+  -- | The @packet@ argument is the packet that will be serialized to the given 
+  -- buffer.
+  Packet ByteString ->
+  m ByteString
+encodeBufferOffPacket ptr unused packet = do
+  (ptr', unused') <- fromPacketBuildR ptr unused packet
+  size <- liftIO (readTotal ptr' unused')
+  liftIO (unsafePackCStringLen ptr' size)
+{-# INLINE encodeBufferOffPacket #-}
+
+withEncodePacket :: 
+  MonadUnliftIO m => 
+  Int -> 
+  ((Packet ByteString -> m ByteString) -> m a) -> 
+  m a
+withEncodePacket n k = 
+  withEncodeBuffer n \ptr -> 
+    k (encodeBufferOffPacket ptr n)
+{-# INLINEABLE withEncodePacket #-}
+
+withEncodeBuffer :: MonadUnliftIO m => Int -> (Ptr Word8 -> m a) -> m a
+withEncodeBuffer packetSizeLimit k = do
+  let bufferSize = metaDataSize + packetSizeLimit
+  buf <- liftIO (newAlignedPinnedByteArray bufferSize metaDataAlign)
+  stateVar <- newIORef BuildRState{currentBuffer = buf, sealedBuffers = mempty}
+  withStablePtr stateVar \statePtr -> do
+    let !p = mutableByteArrayContents buf
+    let !v = plusPtr p bufferSize
+    let !m = plusPtr p metaDataSize
+    liftIO (writeState m statePtr)
+    liftIO (writeSpace m (bufferSize - metaDataSize))
+    k v
+{-# INLINE withEncodeBuffer #-}
+
+-- Unsafe - Helpers ------------------------------------------------------------
+
+-- | Similar to 'bracket', @'withStablePtr' x io@ constructs a managed
+-- 'StablePtr' referring to the value @x@ and provides it to the context that
+-- executes the 'IO' action @io@.
+withStablePtr :: MonadUnliftIO m => a -> (StablePtr a -> m b) -> m b
+withStablePtr x = bracket (liftIO (newStablePtr x)) (liftIO . freeStablePtr)
+
+-- | @unsafePackCStringLen ptr len@ constructs a 'ByteString' from a 'Word8'
+-- array pointer and the length of that array in bytes.
+--
+-- * If the argument to @len@ is less than the actual size of the allocation
+--   @ptr@ refers to, then difference will be dropped from the tail of the array
+--   in a best-case scenario. In the worst-case, the bytestring's foreign ptr
+--   is reclaimed by the garbage collector while it is still alive.
```

Comment:
I find this part confusing.  What does garbage collection semantics have to do with the value of `len`?

Reply:
I see the changes; I'm no longer confused.  Thanks.